### PR TITLE
Capdirty, more vmish

### PR DESCRIPTION
**Pull request by nwf**
A reworking of #405 that might be more aesthetically pleasing in how it glues the MI and MD layers together.  Some XXXes to resolve (probably just eliminate the adjacent code blocks) but more looking for opinions on handling of `PGA_CAPSTORE`.

### CODE REVIEW

**Comment by nwf** [comment]
@bsdjhb asked for an elaboration of the difference relative to #405.  I've pushed rebased revisions to both branches in case anyone wants to look exactly between the two (38468fe557f5 is the current state of #405, and 2dc54a866c49 is the current state of this branch at the time of this writing but I will soon update it to include an adaptation of Brett's Morello work, which is as yet absent on the other branch, too; oof).

The central change is the introduction and use of `sys/vm/vm_page.h` `vm_page_mask_cap_prot`, which strips off `VM_PROT_WRITE_CAP` if the page's `PGA_CAPSTORE` is clear.  While previously the `pmap`s had to infer from `PGA_CAPSTORE` whether `VM_PROT_WRITE_CAP` was safe to interpret, in this respin, all calls to `pmap_enter` that might set that flag are mediated by `vm_page_mask_cap_prot` and the `pmap`s just always do as instructed.  As Jess correctly pointed out, this is probably more in concordance with the existing design, and it shifts some complexity up to the MI bits.

**Comment by nwf** [comment]
CI failure is from unused variables; I've just `#ifdef INVARIANTS` to fix that and can push, but have been holding for other feedback.

File: sys/riscv/riscv/pmap.c [hunk 1]
```diff
@@ -2685,6 +2708,34 @@ pmap_promote_l2(pmap_t pmap, pd_entry_t *l2, vm_offset_t va,
 }
 #endif
 
+#if __has_feature(capabilities)
+static __inline uint64_t
+cheri_pte_cw(vm_page_t m, vm_prot_t prot, u_int flags)
+{
+	vm_page_astate_t mas = vm_page_astate_load(m);
+
+	// XXX Probably not here?
```
Comment (bsdjhb) [inline sys/riscv/riscv/pmap.c:2717]:
```suggestion
	/* XXX Probably not here? */
```

File: sys/riscv/riscv/pmap.c [hunk 2]
```diff
@@ -2685,6 +2708,34 @@ pmap_promote_l2(pmap_t pmap, pd_entry_t *l2, vm_offset_t va,
...
+	KASSERT((mas.flags & PGA_CAPDIRTY) == 0 ||
+	    (mas.flags & PGA_CAPSTORE) != 0,
+	    ("pmap inserting CAPDIRTY w/o CAPSTORE m=%p", m));
+
+	KASSERT((prot & VM_PROT_WRITE_CAP) == 0 ||
+	    (prot & VM_PROT_WRITE) != 0,
+	    ("pmap inserting PROT_WRITE_CAP w/o PROT_WRITE m=%p", m));
+
+	if ((prot & VM_PROT_WRITE_CAP) != 0) {
+		/* XXX PTE_CD from PGA_CAPDIRTY?  Not required... */
```
Comment (bsdjhb) [inline sys/riscv/riscv/pmap.c:2727]: Hmmm, do we want to proactively set `PTE_CD` for `(ftype & VM_PROT_WRITE_CAP) != 0` as we do for `PTE_D` and `VM_PROT_WRITE` faults?

File: sys/riscv/riscv/pmap.c [hunk 3]
```diff
@@ -3483,6 +3531,9 @@ pmap_copy_page(vm_page_t msrc, vm_page_t mdst)
 void
 pmap_copy_page_tags(vm_page_t msrc, vm_page_t mdst)
 {
+
+	VM_PAGE_ASSERT_PGA_CAPMETA_COPY(msrc, mdst);
```
Comment (bsdjhb) [inline sys/riscv/riscv/pmap.c:3535]: FreeBSD's style still doesn't really define variables in the middle of scopes.  I would perhaps move the assertion just before the call to `pagecopy`.

File: sys/vm/vm_fault.c [hunk 4]
```diff
@@ -350,8 +357,20 @@ vm_fault_soft_fast(struct faultstate *fs)
...
 #endif
 	VM_OBJECT_ASSERT_CAP(fs->first_object, fs->prot);
-	rv = pmap_enter(fs->map->pmap, vaddr, m_map,
-	    VM_OBJECT_MASK_CAP_PROT(fs->first_object, fs->prot),
+
+	realprot = VM_OBJECT_MASK_CAP_PROT(fs->first_object, fs->prot);
+
+	/*
+	 * We might be upgrading a page previously mapped VM_PROT_WRITE to one
+	 * now mapped VM_PROT_WRITE|VM_PROT_WRITE_CAP.  Flag it as such.
```
Comment (bsdjhb) [inline sys/vm/vm_fault.c:365]:
```suggestion
	 * now mapped VM_PROT_WRITE | VM_PROT_WRITE_CAP.  Flag it as such.
```

File: sys/vm/vm_fault.c [hunk 5]
```diff
@@ -982,9 +1012,11 @@ vm_fault_cow(struct faultstate *fs)
 		KASSERT(fs->first_object->flags & OBJ_HASCAP,
 		    ("%s: destination object %p doesn't have OBJ_HASCAP",
 		    __func__, fs->first_object));
-		if (fs->object->flags & OBJ_HASCAP)
+		if (fs->object->flags & OBJ_HASCAP) {
+			vm_page_aflag_set(fs->first_m, fs->m->a.flags &
```
Comment (bsdjhb) [inline sys/vm/vm_fault.c:1016]: Thanks, I like doing the vm flags here better.

File: sys/vm/vm_fault.c [hunk 6]
```diff
@@ -2064,9 +2106,12 @@ vm_fault_copy_entry(vm_map_t dst_map, vm_map_t src_map,
 			 * Preserve tags if the source page contains tags.
 			 * See longer discussion in vm_fault_cow.
 			 */
-			if (object->flags & OBJ_HASCAP)
+			if (object->flags & OBJ_HASCAP) {
+				/* Copy across CAPSTORE|CAPDIRTY state, too */
```
Comment (bsdjhb) [inline sys/vm/vm_fault.c:2110]:
```suggestion
				/* Copy across CAPSTORE | CAPDIRTY state, too */
```

File: sys/vm/vm_fault.c [hunk 7]
```diff
@@ -2100,9 +2145,12 @@ vm_fault_copy_entry(vm_map_t dst_map, vm_map_t src_map,
 		 * backing pages.
 		 */
 		if (vm_page_all_valid(dst_m)) {
+			vm_prot_t realprot =
+			    VM_OBJECT_MASK_CAP_PROT(dst_object, prot);
+			realprot = vm_page_mask_cap_prot(dst_m, realprot);
+
 			VM_OBJECT_ASSERT_CAP(dst_object, prot);
```
Comment (bsdjhb) [inline sys/vm/vm_fault.c:2152]: I would perhaps add the `realprot` logic below this assertion so that you validate `prot` first before using it as an input to compute `realprot`.

File: sys/vm/vm_page.h [hunk 8]
```diff
@@ -1011,6 +1073,5 @@ vm_page_domain(vm_page_t m)
 	return (0);
 #endif
 }
-
```
Comment (bsdjhb) [inline sys/vm/vm_page.h:1014]: Seems unrelated?

File: sys/riscv/riscv/pmap.c (hunk 2 above)
Comment (nwf) [reply sys/riscv/riscv/pmap.c:2727]: Good plan; done.

File: sys/arm64/arm64/locore.S [hunk 9]
```diff
@@ -686,7 +686,8 @@ build_l1_block_pagetable:
 	/* Build the L1 block entry */
 	orr	x12, x7, #L1_BLOCK
 #if __has_feature(capabilities)
-	orr	x12, x12, #(ATTR_CAP_RW)
+	orr	x12, x12, #(ATTR_LC_ENABLED)
```
Comment (bsdjhb) [inline sys/arm64/arm64/locore.S:689]: Does this require two instructions or can it be `#(ATTR_LC_ENABLED | ATTR_CDBM)`?

File: sys/arm64/arm64/pmap.c [hunk 10]
```diff
@@ -4893,9 +4937,11 @@ pmap_copy_page(vm_page_t msrc, vm_page_t mdst)
 void
 pmap_copy_page_tags(vm_page_t msrc, vm_page_t mdst)
 {
+
```
Comment (bsdjhb) [inline sys/arm64/arm64/pmap.c:4940]:
```suggestion
```

File: sys/arm64/arm64/pmap.c [hunk 11]
```diff
@@ -709,26 +709,42 @@ pmap_pte_prot(pmap_t pmap, vm_prot_t prot)
...
+	    ("pmap inserting PROT_WRITE_CAP w/o PROT_WRITE m=%p", m));
+
+	if ((prot & VM_PROT_WRITE_CAP) != 0)
+		/*
+		 * The page is CAPSTORE and this mapping is VM_PROT_WRITE_CAP.
+		 * Always set ATTR_CDBM. If this mapping is created in response
+		 * to a cap-write, also set ATTR_SC.
+		 */
+		val |= (ATTR_CDBM);
+
```
Comment (bsdjhb) [inline sys/arm64/arm64/pmap.c:743]:
```suggestion
```

File: sys/arm64/include/pte.h [hunk 12]
```diff
@@ -101,11 +102,10 @@ typedef	uint64_t	pt_entry_t;		/* page table entry */
...
 #if __has_feature(capabilities)
 /*
- * The aarch64 orr instruction cannot handle ATTR_DEFAULT |
- * ATTR_CAP_RW as a single operand, so separate orr instructions are
- * required for ATTR_CAP_RW.
+ * The aarch64 orr instruction cannot handle ATTR_CAP_RW as
+ * a single operand, so separate orr instructions are required.
  */
-#define	ATTR_CAP_RW	(ATTR_LC_ENABLED | ATTR_SC)
+#define	ATTR_CAP_RW	(ATTR_LC_ENABLED | ATTR_CDBM)
```
Comment (bsdjhb) [inline sys/arm64/include/pte.h:108]: Ah, hmm.  Is `ATTR_CAP_RW` used anywhere still after this?  (I guess we might use it in C?)

**Review by bsdjhb — APPROVED** [review]
Sorry, must have missed arm64 last time.  Only a few minor suggestions.

File: sys/arm64/include/pte.h (hunk 12 above)
Comment (nwf) [reply sys/arm64/include/pte.h:108]: Yes, it's used in `pmap.c`.  It's always guarded by `#ifdef __has_feature(capabilities)` and I wonder if it'd be better to guard the definition and make it 0 on non-caps?  But that's minor and can be done later, if it seems like a good plan.

File: sys/vm/vm_page.h (hunk 8 above)
Comment (nwf) [reply sys/vm/vm_page.h:1014]: Arg, this one seems to have sneaked back in. ><

**Comment by nwf** [comment]
I am not sure what I did to drop a brace in arm64 `pmap.c` but I sure did.  Let's try that again.

**Comment by nwf** [comment]
Testing on RISC-V revealed that `UNMANAGED` pages were not being marked capdirty and so could raise fatal faults very early in boot on implementations that did not dirty PTEs natively.

```
diff --git a/sys/riscv/riscv/pmap.c b/sys/riscv/riscv/pmap.c
index 993f312a6c9f..605eea876993 100644
--- a/sys/riscv/riscv/pmap.c
+++ b/sys/riscv/riscv/pmap.c
@@ -2800,6 +2800,8 @@ pmap_enter(pmap_t pmap, vm_offset_t va, vm_page_t m, vm_prot_t prot,
 	if ((m->oflags & VPO_UNMANAGED) != 0) {
 		if (prot & VM_PROT_WRITE)
 			new_l3 |= PTE_D;
+		if (prot & VM_PROT_WRITE_CAP)
+			new_l3 |= PTE_CD;
 	} else
 		new_l3 |= PTE_SW_MANAGED;
 
```

File: sys/arm64/arm64/pmap.c [hunk 13]
```diff
@@ -5189,14 +5233,16 @@ pmap_remove_pages(pmap_t pmap)
 				/*
 				 * Update the vm_page_t clean/reference bits.
 				 */
-				if (pmap_pte_dirty(pmap, tpte)) {
+				if (pmap_pte_dirty(pmap, tpte) ||
+				    pmap_pte_capdirty(pmap, tpte)) {
```
Comment (jrtc27) [inline sys/arm64/arm64/pmap.c:5237]: Does this condition want pushing into the lvl == 1 case?

File: sys/arm64/arm64/trap.c [hunk 14]
```diff
@@ -347,30 +347,28 @@ data_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
...
-		    &ucode);
+		} else
+			ftype = VM_PROT_READ;
+		break;
 	}
+
+	/* Fault in the page. */
+	error = vm_fault_trap(map, far, ftype, VM_FAULT_NORMAL, &sig,
+		&ucode);
+
```
Comment (jrtc27) [inline sys/arm64/arm64/trap.c:371]:
```suggestion
	error = vm_fault_trap(map, far, ftype, VM_FAULT_NORMAL, &sig, &ucode);
```
is what's upstream

File: sys/mips/mips/pmap.c [hunk 15]
```diff
@@ -1561,8 +1561,13 @@ pmap_pv_reclaim(pmap_t locked_pmap)
 				else
 					*pte = 0;
 				m = PHYS_TO_VM_PAGE(TLBLO_PTE_TO_PA(oldpte));
-				if (pte_test(&oldpte, PTE_D))
+				if (pte_test(&oldpte, PTE_D)) {
 					vm_page_dirty(m);
+#ifdef CPU_CHERI
+					if (!pte_test(&oldpte, PTE_SCI))
+						vm_page_capdirty(m);
```
Comment (jrtc27) [inline sys/mips/mips/pmap.c:1568]: Helper function like the other pmaps?

File: sys/mips/mips/pmap.c [hunk 16]
```diff
@@ -2238,7 +2260,13 @@ pmap_enter(pmap_t pmap, vm_offset_t va, vm_page_t m, vm_prot_t prot,
 	origpte = *pte;
 	KASSERT(!pte_test(&origpte, PTE_D | PTE_RO | PTE_V),
 	    ("pmap_enter: modified page not writable: va: %p, pte: %#jx",
-	    (void *)(uintptr_t)va, (uintmax_t)origpte));
+	    (void *)va, (uintmax_t)origpte));
```
Comment (jrtc27) [inline sys/mips/mips/pmap.c:2263]: Why?

File: sys/riscv/riscv/pmap.c [hunk 17]
```diff
@@ -2731,8 +2785,7 @@ pmap_enter(pmap_t pmap, vm_offset_t va, vm_page_t m, vm_prot_t prot,
 #if __has_feature(capabilities)
 	if (prot & VM_PROT_READ_CAP)
 		new_l3 |= PTE_CR;
-	if (prot & VM_PROT_WRITE_CAP)
-		new_l3 |= PTE_CW | PTE_CD;
+	new_l3 |= cheri_pte_cw(m, prot, flags);
```
Comment (jrtc27) [inline sys/riscv/riscv/pmap.c:2788]:
```suggestion
	if (flags & VM_PROT_WRITE_CAP)
		new_l3 |= PTE_CD;
	if (prot & VM_PROT_WRITE_CAP)
		new_l3 |= PTE_CW;
```
to mirror D/W above, no need to introduce a function for that

File: sys/riscv/riscv/pmap.c [hunk 18]
```diff
@@ -3483,9 +3538,11 @@ pmap_copy_page(vm_page_t msrc, vm_page_t mdst)
 void
 pmap_copy_page_tags(vm_page_t msrc, vm_page_t mdst)
 {
+
```
Comment (jrtc27) [inline sys/riscv/riscv/pmap.c:3541]:
```suggestion
```

File: sys/arm64/arm64/pmap.c (hunk 13 above)
Comment (brettferdosi) [reply sys/arm64/arm64/pmap.c:5221]: Happy to do this, it would save an extra check in the non-block case.

**Comment by nwf** [comment]
This push picks up a fix to define away `VM_PAGE_ASSERT_PGA_CAPMETA_PMAP_ENTER` if `!INVARIANTS` and @brettferdosi's optimization of `pmap_remove_pages`.  Let's see how CI feels about it.

File: sys/arm64/include/pte.h (hunk 12 above)
Comment (bsdjhb) [reply sys/arm64/include/pte.h:108]: I think I prefer keeping it undefined for !capabilities and explicit #ifdef's for its use as is currently done.

File: sys/arm64/arm64/pmap.c [hunk 19]
```diff
@@ -5485,9 +5514,7 @@ pmap_remove_write(vm_page_t m)
 			if (!atomic_fcmpset_long(pte, &oldpte,
 			    (oldpte | ATTR_S1_AP_RW_BIT) & ~ATTR_SW_DBM))
 				goto retry;
-			if ((oldpte & ATTR_S1_AP_RW_BIT) ==
-			    ATTR_S1_AP(ATTR_S1_AP_RW))
-				vm_page_dirty(m);
```
Comment (jrtc27) [inline sys/arm64/arm64/pmap.c:5490]: Note to self/other reviewers: this didn't previously check DBM but pmap_page_dirty does (pmap_pte_dirty, called by pmap_pte_dirty, for S1 mappings, is checking DBM+RW). This is because this code was previously micro-optimised due to already being in the DBM != 0 case, i.e. manually inlined pmap_pte_dirty and removed the redundant condition. So this change is fine (and hopefully the compiler will inline and optimise some parts accordingly).

File: sys/arm64/arm64/pmap.c [hunk 20]
```diff
@@ -5771,7 +5795,7 @@ pmap_advise(pmap_t pmap, vm_offset_t sva, vm_offset_t eva, int advice)
 					 * dirty now.
 					 */
 					m = PHYS_TO_VM_PAGE(oldl3 & ~ATTR_MASK);
-					vm_page_dirty(m);
+					pmap_page_dirty(pmap, oldl3, m);
```
Comment (jrtc27) [inline sys/arm64/arm64/pmap.c:5798]: This seems suspicious; it's under a pmap_pte_dirty check above, no pmap_pte_capdirty check.

File: sys/arm64/arm64/trap.c [hunk 21]
```diff
@@ -347,30 +347,27 @@ data_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
...
+			ftype = VM_PROT_WRITE;
 #if __has_feature(capabilities)
-	if ((esr & ISS_DATA_DFSC_MASK) == ISS_DATA_DFSC_LC_SC) {
-		sig = SIGSEGV;
-		ucode = (esr & ISS_DATA_WnR) == 0 ? SEGV_LOADTAG :
-		    SEGV_STORETAG;
-		error = KERN_FAILURE;
-	} else
+			if ((esr & ISS_DATA_DFSC_MASK) ==
+				ISS_DATA_DFSC_LC_SC)
```
Comment (jrtc27) [inline sys/arm64/arm64/trap.c:360]: This should fit just fine on one line (78 < 80), and if you were to wrap it, it should only have 4 spaces for the continuation line

File: sys/arm64/arm64/trap.c [hunk 22]
```diff
@@ -347,30 +347,27 @@ data_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
...
-		error = vm_fault_trap(map, far, ftype, VM_FAULT_NORMAL, &sig,
-		    &ucode);
+		} else
+			ftype = VM_PROT_READ;
+		break;
 	}
+
+	/* Fault in the page. */
+	error = vm_fault_trap(map, far, ftype, VM_FAULT_NORMAL, &sig, &ucode);
+
```
Comment (jrtc27) [inline sys/arm64/arm64/trap.c:370]: Remove this blank line, it's not present upstream

File: sys/arm64/include/pte.h (hunk 12 above)
Comment (jrtc27) [reply sys/arm64/include/pte.h:108]: 0 for !capabilities doesn't help you in assembly, since `0` (and `-1`) are not immediates that can be encoded for `orr`, I guess for obvious reasons.

File: sys/mips/mips/trap.c [hunk 23]
```diff
@@ -691,6 +691,49 @@ cpu_fetch_syscall_args(struct thread *td)
...
+		if (KERNLAND(va)) {
+			return (is_kernel &&
+			    !pmap_emulate_capdirty(kernel_pmap, va));
+		} else {
+			pmap_t upmap = vmspace_pmap(uvms);
+			return (!pmap_emulate_capdirty(upmap, va));
+		}
+	}
+
+	return false;
```
Comment (jrtc27) [inline sys/mips/mips/trap.c:734]:
```suggestion
	return (false);
```

File: sys/mips/mips/trap.c [hunk 24]
```diff
@@ -1162,6 +1205,19 @@ trap(struct trapframe *trapframe)
...
+		if (!c2e_fixup_fault(trapframe, true, p->p_vmspace, &ftype)) {
+			if (ftype != 0) {
+				if (KERNLAND(trapframe->badvaddr))
+					goto kernel_fault;
+				else
+					goto dofault;
+			}
+		} else {
+			return (trapframe->pc);
+		}
```
Comment (jrtc27) [inline sys/mips/mips/trap.c:1219]: These would make more sense to me if you flipped the order and hoisted the slow path out below, ie:

```
if (c2_fixup_fault(...))
	return (trapframe->pc);

if (ftype != 0) {
	...
}
```

File: sys/vm/vm_fault.c [hunk 25]
```diff
@@ -664,6 +687,13 @@ vm_fault_trap(vm_map_t map, vm_offset_t vaddr, vm_prot_t fault_type,
 			*ucode = BUS_OBJERR;
 			break;
 		case KERN_PROTECTION_FAILURE:
+#if __has_feature(capabilities)
+			if ((fault_type & VM_PROT_WRITE_CAP) != 0)
+				segv_ucode = SEGV_STORETAG;
+			else
+#endif
```
Comment (jrtc27) [inline sys/vm/vm_fault.c:694]: Do we need to bother with #if here? We don't in various places elsewhere. It's dead code, but it's a slow path, and should compile just fine.

**Comment by nwf** [comment]
@brettferdosi Could you prep an additional commit, or amend existing commits, in light of @jrtc27's feedback?  I think you should be able to `push -f` over the branch here, but if not point me at another branch to pull and I'll do the reconciliation.

File: sys/arm64/arm64/pmap.c [hunk 26]
```diff
@@ -709,26 +709,25 @@ pmap_pte_prot(pmap_t pmap, vm_prot_t prot)
...
 			val |= ATTR_S2_XN(ATTR_S2_XN_ALL);
+	}
+
 #if __has_feature(capabilities)
-		if ((prot & VM_PROT_READ_CAP) != 0)
-			val |= ATTR_LC_ENABLED;
-		if ((prot & VM_PROT_WRITE_CAP) != 0)
-			val |= ATTR_SC;
+	if ((prot & VM_PROT_READ_CAP) != 0)
+		val |= ATTR_LC_ENABLED;
```
[the remark this answers is not in the document]
Comment (brettferdosi) [reply sys/arm64/arm64/pmap.c:723]: Yeah, I thought this case was fine because it's the same bit for stage 1 and 2.

File: sys/arm64/arm64/pmap.c (hunk 20 above)
Comment (brettferdosi) [reply sys/arm64/arm64/pmap.c:5798]: pmap_page_dirty will do the pmap_pte_capdirty check before capdirtying the vm_page; I thought it might be beneficial to propagate capdirty state in addition to normal dirty state to the page at this point. Happy to revert if you think that would be better.

File: sys/arm64/arm64/pmap.c [hunk 27]
```diff
@@ -709,26 +709,25 @@ pmap_pte_prot(pmap_t pmap, vm_prot_t prot)
...
+
+	VM_PAGE_ASSERT_PGA_CAPMETA_PMAP_ENTER(m, prot);
+	if ((prot & VM_PROT_WRITE_CAP) != 0)
+		val |= ATTR_CDBM;
+	if ((flags & VM_PROT_WRITE_CAP) != 0)
+		val |= ATTR_SC;
```
[the remark this answers is not in the document]
Comment (brettferdosi) [reply sys/arm64/arm64/pmap.c:729]: The SC bit was included to match the behavior of the riscv pmap_enter: if the flags argument to that function has VM_PROT_WRITE_CAP, then the pte has the capdirty bit set.

As for CDBM, the corresponding code that sets DBM is further down in pmap_enter: line 3925 in this diff. I thought it would make more sense to keep the capability related stuff together (though I've just lifted the assert out of pmap_pte_prot and into pmap_enter in line with your above comment).

**Comment by nwf** [comment]
Ah, right, this... This isn't going to work correctly as written (not to mention that this PR is quite stale).  See https://github.com/CTSRD-CHERI/paper-revocation3/issues/24 for discussion and once things are flying in their current broken state again I'll take another stab at fixing this.
